### PR TITLE
fix table version determination for psu init and jobinstanceentityextractor to choose correct version

### PR DESCRIPTION
**Pull request by ajm01**
addresses WAS RTC defect 250673

### CODE REVIEW

**Comment by ajm01** [comment]
#build

File: dev/com.ibm.jbatch.container/src/com/ibm/jbatch/container/services/impl/JPAPersistenceManagerImpl.java [hunk 1]
```diff
@@ -2593,12 +2595,15 @@ protected void rollbackIfNewTranWasStarted(Throwable caughtThrowable) throws Per
             }
 
             // If we haven't gotten a new exception to chain, throw back the original one passed in as a parameter
+            logger.fine("No exception encountered, throwing new PersistenceException with original exception chained");
             throw new PersistenceException(caughtThrowable);
         }
 
         protected void resumeAnyExistingLTC() {
+            logger.fine("Will resume any LTC");
```
Comment (scottkurz) [inline dev/com.ibm.jbatch.container/src/com/ibm/jbatch/container/services/impl/JPAPersistenceManagerImpl.java]: not needed.. entry/exit is enough here.  Don't care either way about the next one.

File: dev/com.ibm.jbatch.container/src/com/ibm/jbatch/container/services/impl/JPAPersistenceManagerImpl.java [hunk 2]
```diff
@@ -2712,7 +2717,8 @@ public Integer call() throws Exception {
             }.runInNewOrExistingGlobalTran();
 
             return exec;
-        } catch (javax.persistence.PersistenceException pe) {
+        } catch (PersistenceException pe) {
+            logger.fine("Caught a persistence exception");
             Throwable cause = pe.getCause();
             while (cause != null) {
```
Comment (scottkurz) [inline dev/com.ibm.jbatch.container/src/com/ibm/jbatch/container/services/impl/JPAPersistenceManagerImpl.java]: Let's trace each cause along the chain, printing the exc itself and the exc.getMessage either here or at the end of the loop if we haven't returned.     THis way if we step through a chain and don't match we have evidence along the way

File: dev/com.ibm.jbatch.container/src/com/ibm/jbatch/container/services/impl/JPAPersistenceManagerImpl.java [hunk 3]
```diff
@@ -2683,18 +2688,18 @@ public int getJobInstanceTableVersion() throws Exception {
...
 
+        // Check for the presence of v2 table support (presence of UPDATETIME column)
         try {
             Integer exec = new TranRequest<Integer>(em) {
                 @Override
                 public Integer call() throws Exception {
-                    // Verify that UPDATETIME column exists by running a query against it.
                     String queryString = "SELECT COUNT(x.lastUpdatedTime) FROM JobInstanceEntityV2 x";
                     TypedQuery<Long> query = em.createQuery(queryString, Long.class);
                     query.getSingleResult();
```
Comment (scottkurz) [inline dev/com.ibm.jbatch.container/src/com/ibm/jbatch/container/services/impl/JPAPersistenceManagerImpl.java]: Let's break this up into two separate try-catch blocks... one to see if we can make it to V2.. the next to V3.

**Comment by ajm01** [comment]
#build

**Comment by ajm01** [comment]
#build

**Comment by scottkurz** [comment]
Made some comments in a commit (on top of yours) here:
https://github.com/scottkurz/open-liberty/commit/388e25a70f7f690becef80daa48ab5e6700db28c?w=1

**Comment by scottkurz** [comment]
@ajm01  my eyes are glazing over.   I'd say go ahead and merge if/when the build completes OK, but maybe @cgianfra  you could take one more look when you get a chance.  Thanks

**Comment by scottkurz** [comment]
Kicked off new build at:
https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_QEYJQA5-EeiStsw2He7iFw

Earlier failed w/ infra issues, (plus doesn't look like it had 2nd commit from me).

**Comment by ajm01** [comment]
#build

**Comment by scottkurz** [comment]
This last build hits a real failure (in batch):  ` batch.fat.junit.LocalServerJobRecoveryAtStartUpTest.recoverLocalJobsInStartingTest`

There might be an infinite loop in:

	at com.ibm.jbatch.container.ws.impl.WSStartupRecoveryServiceImpl.recoverLocalJobsInInflightStates(WSStartupRecoveryServiceImpl.java:126)
	at com.ibm.jbatch.container.services.impl.JPAPersistenceManagerImpl.createPsu(JPAPersistenceManagerImpl.java:347)
	at com.ibm.jbatch.container.services.impl.JPAPersistenceManagerImpl.getPsu(JPAPersistenceManagerImpl.java:259)
	at com.ibm.jbatch.container.services.impl.JPAPersistenceManagerImpl.getJobExecutionTableVersion(JPAPersistenceManagerImpl.java:2636)
	at com.ibm.jbatch.container.persistence.jpa.JobExecutionEntityExtractor.extractClassFromRow(JobExecutionEntityExtractor.java:39)
	at org.eclipse.persistence.descriptors.InheritancePolicy.classFromRow(InheritancePolicy.java:343)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObject(ObjectBuilder.java:725)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObject(ObjectBuilder.java:691)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.buildObject(ObjectLevelReadQuery.java:805)
	at org.eclipse.persistence.queries.ReadAllQuery.registerResultInUnitOfWork(ReadAllQuery.java:962)
	at org.eclipse.persistence.queries.ReadAllQuery.executeObjectLevelReadQuery(ReadAllQuery.java:573)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeDatabaseQuery(ObjectLevelReadQuery.java:1175)
	at org.eclipse.persistence.queries.DatabaseQuery.execute(DatabaseQuery.java:911)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.execute(ObjectLevelReadQuery.java:1134)
	at org.eclipse.persistence.queries.ReadAllQuery.execute(ReadAllQuery.java:460)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeInUnitOfWork(ObjectLevelReadQuery.java:1222)
	at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.internalExecuteQuery(UnitOfWorkImpl.java:2899)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1863)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1845)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1810)
	at org.eclipse.persistence.internal.jpa.QueryImpl.executeReadQuery(QueryImpl.java:258)
	at org.eclipse.persistence.internal.jpa.QueryImpl.getResultList(QueryImpl.java:473)
	at com.ibm.jbatch.container.services.impl.JPAPersistenceManagerImpl$27.call(JPAPersistenceManagerImpl.java:1215)
	at com.ibm.jbatch.container.services.impl.JPAPersistenceManagerImpl$27.call(JPAPersistenceManagerImpl.java:1208)
	at com.ibm.jbatch.container.services.impl.JPAPersistenceManagerImpl$TranRequest.runInNewOrExistingGlobalTran(JPAPersistenceManagerImpl.java:2520)
	at com.ibm.jbatch.container.services.impl.JPAPersistenceManagerImpl.getJobExecutionsRunningLocalToServer(JPAPersistenceManagerImpl.java:1221)
	at com.ibm.jbatch.container.ws.impl.WSStartupRecoveryServiceImpl.recoverLocalJobsInInflightStates(WSStartupRecoveryServiceImpl.java:126)
	at com.ibm.jbatch.container.services.impl.JPAPersistenceManagerImpl.createPsu(JPAPersistenceManagerImpl.java:347)
	at com.ibm.jbatch.container.services.impl.JPAPersistenceManagerImpl.getPsu(JPAPersistenceManagerImpl.java:259)
	at com.ibm.jbatch.container.services.impl.JPAPersistenceManagerImpl.getJobExecution(JPAPersistenceManagerImpl.java:1115)
	at com.ibm.jbatch.container.api.impl.JobOperatorImpl.getJobExecution(JobOperatorImpl.java:398

**Comment by ajm01** [comment]
#build

**Comment by scottkurz** [comment]
We can close this since the other #2143  was merged.

**Pull request closed without merging**